### PR TITLE
Postgres schema acceptance process scripts

### PR DESCRIPTION
**Pull request by Will-Howard**
Add `yarn makemigrations` and `yarn acceptmigrations` scripts for managing schema changes.

It should be possible for people to understand this by reading README.md and grepping around the codebase so I won't redescribe it here. If you find it hard to follow from the code then tell me to add better documentation

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203212013479101) by [Unito](https://www.unito.io)

### CODE REVIEW

File: schema/accepted_schema.sql [hunk 1]
```diff
@@ -0,0 +1,910 @@
...
+    "canonicalSource" text,
+    "nominationCount2018" real DEFAULT 0,
+    "nominationCount2019" real DEFAULT 0,
+    "reviewCount2018" real DEFAULT 0,
+    "reviewCount2019" real DEFAULT 0,
+    "reviewCount" real DEFAULT 0,
+    "reviewVoteCount" real DEFAULT 0,
+    "positiveReviewVoteCount" real DEFAULT 0,
+    "reviewVoteScoreAF" real DEFAULT 0,
+    "reviewVotesAF" real[] DEFAULT '{}' ::real[][],
```
Comment (Will-Howard) [inline schema/accepted_schema.sql:386]: Should be:
```suggestion
    "reviewVotesAF" real[] DEFAULT '{}' ::real[],
```
There is a bug in the default generating code, which I've fixed in this PR: https://github.com/ForumMagnum/ForumMagnum/pull/5926

File: packages/lesswrong/integrationTests/schemacheck.test.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import "./integrationTestSetup";
+import path from "path";
+import { makeMigrations } from "../server/scripts/makeMigrations";
+
+describe('Schema check', () => {
+  it('Has an accepted_schema.sql file which matches the schema defined in code', async () => {
+    await makeMigrations({writeSchemaChangelog: false, writeAcceptedSchema: false, generateMigrations: false, rootPath: path.join(__dirname, "../../../")});
```
Comment (Will-Howard) [inline packages/lesswrong/integrationTests/schemacheck.test.ts:7]: It might make sense to split this out into it's own CI stage at some point so it runs faster (it will be annoying to wait for all the tests to pass only to find that you forgot to run `yarn acceptschema`), but let's wait and see how much of a problem this is

**Review by oetherington — APPROVED** [review]
Nice work! I left a few comments but there's nothing too major.

File: packages/lesswrong/integrationTests/voting.tests.ts [hunk 3]
```diff
@@ -63,9 +63,9 @@ describe('Voting', function() {
         Posts.findOne({_id: curatedPost._id}),
       ]);
 
-      (updatedNormalPost?.score as any).should.be.closeTo(recalculateScore(normalPost), 0.001);
-      (updatedFrontpagePost?.score as any).should.be.closeTo(recalculateScore(frontpagePost), 0.001);
-      (updatedCuratedPost?.score as any).should.be.closeTo(recalculateScore(curatedPost), 0.001);
+      (updatedNormalPost?.score as any).should.be.closeTo(recalculateScore(normalPost), 0.002);
+      (updatedFrontpagePost?.score as any).should.be.closeTo(recalculateScore(frontpagePost), 0.002);
+      (updatedCuratedPost?.score as any).should.be.closeTo(recalculateScore(curatedPost), 0.002);
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/integrationTests/voting.tests.ts:68]: 🥲

File: scripts/makeMigrations.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Entry point for `yarn makemigrations`
+ *
+ * This function is a simple wrapper around the `makeMigrations` script to provide some
+ * feedback to the user (because it runs as a server script any log output it sent to the server stdout).
+ * See `packages/lesswrong/server/scripts/makeMigrations.ts` for more complete documentation.
+ * @returns
+ */
+const run = async () => {
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply scripts/makeMigrations.ts:15]: 👍 `scripts/serverShellCommand.sh` already checks for this so I'll just add an `echo` to say so

File: packages/lesswrong/server/scripts/makeMigrations.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+ *   It contains a list of hashes of schema files that have been accepted. If the current schema hash is the most recent entry in this file, then the schema is accepted.
+ * - `accepted_schema.sql`: This is a SQL view of the schema that has been accepted.
+ * - `schema_to_accept.sql`: If the current schema is not accepted, this file will be generated to contain a SQL view of the "unaccepted" schema.
+ *   This is useful for comparing against the accepted schema to see what changes need to be made in the migration that is generated. It is automatically deleted when the schema is accepted.
+ *
+ * @param {boolean} writeSchemaChangelog - If true, update the schema_changelog.json file before checking for changes
+ * @param {boolean} writeAcceptedSchema - If true, update the `accepted_schema.sql` and `schema_to_accept.sql`
+ * @param {boolean} generateMigrations - If true, generate a template migration file if the schema has changed
+ * @param {boolean} rootPath - The root path of the project, this is annoying but required because this script is sometimes run from the server bundle, and sometimes from a test.
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/scripts/makeMigrations.ts:103]: Hmm, yeah I think I would prefer that, although it's still not ideal

File: packages/lesswrong/lib/collections/users/schema.ts [hunk 6]
```diff
@@ -749,7 +749,8 @@ const schema: SchemaType<DbUser> = {
     canRead: userOwns,
     canUpdate: [userOwns, 'sunshineRegiment', 'admins'],
     canCreate: 'guests',
-    ...schemaDefaultValue(getDefaultFilterSettings),
+    // FIXME this isn't filling default values as intended
+    // ...schemaDefaultValue(getDefaultFilterSettings),
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/lib/collections/users/schema.ts:753]: It's not working on master either, we don't do anything clever to check if the default value is a function and fill it in

File: scripts/acceptMigrations.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import { fs } from "mz";
+
+const run = async () => {
+  await acceptMigrations({write: true, rootPath: path.join(__dirname, "../")});
+
+  const runExec = promisify(exec)
+  await runExec("./scripts/serverShellCommand.sh --wait \"Vulcan.makeMigrations({generateMigrations: false, writeSchemaChangelog: false})\"")
+
+  if (fs.existsSync('./schema/schema_to_accept.sql')) {
+    console.error("Error: Not all schema changes have been accepted.");
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply scripts/acceptMigrations.ts:14]: I'll show the `git diff` here

File: packages/lesswrong/integrationTests/schemacheck.test.ts (hunk 2 above)
Comment (Will-Howard) [reply packages/lesswrong/integrationTests/schemacheck.test.ts:7]: Good shout! This is another thing I tried and failed to get to work before but I have now managed it by doing `require('../server.ts');`
